File: Java/src/main/java/com/gildedrose/Item.java
package com.gildedrose;

public class Item {

    public String name;

    public int sellIn;

    public int quality;

    public Item(String name, int sellIn, int quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }

    @Override
    public String toString() {
        return this.name + ", " + this.sellIn + ", " + this.quality;
    }

    void updateItem() {
        if ("Aged Brie".equals(name)) {
            increaseQualityIfBelow50();
        } else if ("Backstage passes to a TAFKAL80ETC concert".equals(name)) {
            increaseQualityIfBelow50();

            if (sellIn < 11) {
                increaseQualityIfBelow50();
            }

            if (sellIn < 6) {
                increaseQualityIfBelow50();
            }
        } else if ("Sulfuras, Hand of Ragnaros".equals(name)) {
        } else {
            if (quality > 0) {
                quality = quality - 1;
            }
        }

        if ("Sulfuras, Hand of Ragnaros".equals(name)) {
        } else {
            sellIn = sellIn - 1;
        }

        if ("Aged Brie".equals(name)) {
            if (sellIn < 0) {
                increaseQualityIfBelow50();
            }
        } else if ("Backstage passes to a TAFKAL80ETC concert".equals(name)) {
            if (sellIn < 0) {
                quality = quality - quality;
            }
        } else if ("Sulfuras, Hand of Ragnaros".equals(name)) {
        } else {
            if (sellIn < 0) {
                if (quality > 0) {
                    quality = quality - 1;
                }
            }
        }
    }

    private void increaseQualityIfBelow50() {
        if (quality < 50) {
            quality = quality + 1;
        }
    }
}

File: scala/src/main/scala/com/gildedrose/Item.scala
package com.gildedrose

object Factory {
  def createItem(name: String, sellIn: Int, quality: Int): Item = {
    name match {
      case "Aged Brie" => new AgedBrieItem(name, sellIn, quality)
      case "Backstage passes to a TAFKAL80ETC concert" => new BackstageItem(name, sellIn, quality)
      case "Sulfuras, Hand of Ragnaros" => new SulfurasItem(name, sellIn, quality)
      case _ => new Item(name, sellIn, quality)
    }
  }
}

class Item(val name: String, var sellIn: Int, var quality: Int) {

  final def updateItem(): Unit = {
    updateQuality()
    updateExpiration()
    if (isExpired) {
      updateAfterExpiration()
    }
  }

  protected def updateAfterExpiration(): Unit = {
    decreaseQuality()
  }

  protected def updateExpiration(): Unit = {
    decreaseSellIn()
  }

  protected def updateQuality(): Unit = {
    decreaseQuality()
  }

  final protected def isExpired: Boolean = sellIn < 0

  final protected def decreaseSellIn(): Unit = {
    sellIn = sellIn - 1
  }

  final protected def decreaseQuality(): Unit = {
    if (quality > 0) {
      quality = quality - 1
    }
  }

  final protected def dropQuality(): Unit = {
    quality = 0
  }

  final protected def increaseQuality(): Unit = {
    if (quality < 50) {
      quality = quality + 1
    }
  }
}